crates/sui-framework/sources/transfer.move

```diff
@@ -19,15 +19,46 @@ module sui::transfer {
     /// longer be transferred or mutated.
     public native fun freeze_object<T: key>(obj: T);
 
+    spec freeze_object {
+        pragma opaque;
+        // never aborts as it requires object by-value and:
+        // - it's OK to freeze whether object is fresh or owned
+        // - shared or immutable object cannot be passed by value
+        aborts_if [abstract] false;
+        modifies [abstract] global<sui::object::Ownership>(sui::object::id(obj).bytes);
+        ensures [abstract] exists<sui::object::Ownership>(sui::object::id(obj).bytes);
+        ensures [abstract] global<sui::object::Ownership>(sui::object::id(obj).bytes).status == 3 /* IMMUTABLE */;
```

Comment:
I believe we are actually able to use constants from other modules in specs (unless things have changed) so you might be able to use `sui::prover::IMMUTABLE` here, and same for `SHARED` and `OWNED`.